### PR TITLE
Fix book editing issues

### PR DESCRIPTION
**Pull request by barbaravaldez**
- Fix issue moving a duplicated section to another book. Before it wouldn't remove the section from the source toc.yml, because it was looking for the duplicated name and path instead of the original values.
- Add additional check for verifying that the file exists in the new path before deleting the copy.
- Initialize the movedFiles, modifiedDirectories and tocFiles maps every time we try to edit a book.
- Remove setFileWatcher on the targetBook toc, because we never unwatched the file.

### CODE REVIEW

File: extensions/notebook/src/book/bookTocManager.ts [hunk 1]
```diff
@@ -138,11 +141,15 @@ export class BookTocManager implements IBookTocManager {
 			contents.forEach(async (content) => {
 				if ((await fs.stat(path.join(directory, content))).isFile) {
```
Comment (corivera) [inline extensions/notebook/src/book/bookTocManager.ts:142]: You do this same path.join() call a couple times in this method, so you can probably cache it in a single variable.

File: extensions/notebook/src/book/bookTocManager.ts [hunk 2]
```diff
@@ -373,11 +380,11 @@ export class BookTocManager implements IBookTocManager {
 			}
 		}
 		else if (element.contextValue === 'savedNotebook') {
+			// the notebook is part of a book so we need to modify its toc as well
+			const findSection = { file: element.book.page?.file?.replace(/\\/g, '/'), title: element.book.page?.title };
```
Comment (corivera) [inline extensions/notebook/src/book/bookTocManager.ts:385]: Why did this get moved? Looks like it's only used in the previous location.

Comment (barbaravaldez) [reply extensions/notebook/src/book/bookTocManager.ts:385]: The name of the file is changed if there is a file with the same name in the target book in the addNotebook method. In the source book toc, I need to  look for the notebook using it's unchanged name in order to remove it.

**Review by corivera — APPROVED** [review]

File: extensions/notebook/src/book/bookTocManager.ts [hunk 3]
```diff
@@ -44,6 +44,9 @@ export class BookTocManager implements IBookTocManager {
 		this._sourceBook = sourceBook;
 		this.newSection = {};
 		this.tableofContents = [];
+		this._movedFiles = new Map<string, string>();
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/book/bookTocManager.ts:47]: Why did you move this? If you just initialize like you had before that's doing the exact same thing

File: extensions/notebook/src/book/bookTocManager.ts [hunk 4]
```diff
@@ -136,16 +139,21 @@ export class BookTocManager implements IBookTocManager {
...
+						// if the file exists in the new path and if the the new path and old path are different
+						// then we can remove it
+						if (exists && newPath !== filePath) {
+							// the file could not be renamed, so a copy was created.
+							await fs.unlink(filePath);
+						}
 					}
-				} else if ((await fs.stat(path.join(directory, content))).isDirectory) {
-					await this.cleanUp(path.join(directory, content));
+				} else if ((await fs.stat(filePath)).isDirectory) {
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/book/bookTocManager.ts:155]: You should just call stat once at the start

File: extensions/notebook/src/book/bookTocManager.ts [hunk 5]
```diff
@@ -354,10 +362,10 @@ export class BookTocManager implements IBookTocManager {
 	public async updateBook(element: BookTreeItem, targetBook: BookTreeItem, targetSection?: JupyterBookSection): Promise<void> {
 		const targetBookVersion = targetBook.book.version === BookVersion.v1 ? BookVersion.v1 : BookVersion.v2;
 		if (element.contextValue === 'section') {
-			await this.addSection(element, targetBook);
-			const elementVersion = element.book.version === BookVersion.v1 ? BookVersion.v1 : BookVersion.v2;
 			// modify the sourceBook toc and remove the section
 			const findSection: JupyterBookSection = { file: element.book.page.file?.replace(/\\/g, '/'), title: element.book.page.title };
+			await this.addSection(element, targetBook);
+			const elementVersion = element.book.version === BookVersion.v1 ? BookVersion.v1 : BookVersion.v2;
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/book/bookTocManager.ts:368]: This can only be v1 or v2 so not sure what this check is doing. Could just be this right?

`const elementVersion = element.book.version;`

File: extensions/notebook/src/book/bookTocManager.ts [hunk 6]
```diff
@@ -373,11 +381,11 @@ export class BookTocManager implements IBookTocManager {
 			}
 		}
 		else if (element.contextValue === 'savedNotebook') {
+			// the notebook is part of a book so we need to modify its toc as well
+			const findSection = { file: element.book.page?.file?.replace(/\\/g, '/'), title: element.book.page?.title };
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/book/bookTocManager.ts:385]: It seems like this is the kind of thing that should probably go in a more common place such as updateTOC so that you're always sure you're writing out valid paths and every caller doesn't have to do this themselves.

File: extensions/notebook/src/book/bookTreeView.ts [hunk 7]
```diff
@@ -213,9 +213,9 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
 			const sourceBook = this.books.find(book => book.bookPath === movingElement.book.root);
 			const targetBook = this.books.find(book => book.bookPath === updateBook.book.root);
 			this.bookTocManager = new BookTocManager(targetBook, sourceBook);
-			// remove watch on toc file from both books.
+			// remove watch on toc file from source book.
 			if (sourceBook) {
-				fs.unwatchFile(movingElement.tableOfContentsPath);
+				fs.unwatchFile(sourceBook.tableOfContentsPath);
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/book/bookTreeView.ts:218]: Having it call setFileWatcher with just sourceBook but then needing to know that it's actually watching tableOfContentsPath for the unwatch isn't really intuitive. What if someone changes setFileWatcher to watch a different file? There's no way to know that this is just going to stop doing anything unless the person happens to know/catch it.

What would be really cool is if you had setFileWatcher return a disposable - that when disposed would automatically unhook the file watcher. That would be a lot safer since all the logic for hooking/unhooking would be done in the same place. 

See https://github.com/microsoft/azuredatastudio/blob/main/extensions/azdata/src/azdata.ts#L38 for a simple example of doing this - the class implements vscode.Disposable and then takes care of cleaning up the session when it's disposed.

**Review by Charles-Gagnon — APPROVED** [review]

File: extensions/notebook/src/book/bookTocManager.ts (hunk 3 above)
Comment (Charles-Gagnon) [reply extensions/notebook/src/book/bookTocManager.ts:47]: I'm assuming this is what your statement is about : 

`Initialize the movedFiles, modifiedDirectories and tocFiles maps every time we try to edit a book.`

But this is only going to do that when we construct a new BookTocManager...at which point what you had before is also going to be doing that.

File: extensions/notebook/src/book/bookTreeView.ts (hunk 7 above)
Comment (barbaravaldez) [reply extensions/notebook/src/book/bookTreeView.ts:218]: I'll implement this next week, since I'm going to be adding more operations to the bookTocManager

File: extensions/notebook/src/book/bookTocManager.ts (hunk 5 above)
Comment (barbaravaldez) [reply extensions/notebook/src/book/bookTocManager.ts]: Yup, I just need to cast it so it's a BookVersion object.

File: extensions/notebook/src/book/bookTocManager.ts (hunk 6 above)
Comment (barbaravaldez) [reply extensions/notebook/src/book/bookTocManager.ts:386]: I'll be correcting this on my next PR.
